src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/Endpoints/SemanticTokens.cs

```diff
@@ -44,35 +48,48 @@ internal partial class RazorCustomMessageTarget
             return new ProvideSemanticTokensResponse(tokens: null, hostDocumentSyncVersion: -1);
         }
 
-        semanticTokensParams.TextDocument.Uri = csharpDoc.Uri;
-
-        var newParams = new SemanticTokensRangeParams
-        {
-            TextDocument = semanticTokensParams.TextDocument,
-            PartialResultToken = semanticTokensParams.PartialResultToken,
-            Range = semanticTokensParams.Range,
-        };
+        // Ensure the C# ranges are sorted
+        Array.Sort(semanticTokensParams!.Ranges, static (r1, r2) => r1.CompareTo(r2));
+        var requestTasks = new List<Task<ReinvocationResponse<VSSemanticTokensResponse>?>>(semanticTokensParams!.Ranges.Length);
 
+        semanticTokensParams.TextDocument.Uri = csharpDoc.Uri;
         var textBuffer = csharpDoc.Snapshot.TextBuffer;
         var languageServerName = RazorLSPConstants.RazorCSharpLanguageServerName;
         var lspMethodName = Methods.TextDocumentSemanticTokensRangeName;
-        using var _ = _telemetryReporter.TrackLspRequest(lspMethodName, languageServerName, semanticTokensParams.CorrelationId);
-        var csharpResults = await _requestInvoker.ReinvokeRequestOnServerAsync<SemanticTokensRangeParams, VSSemanticTokensResponse>(
-            textBuffer,
-            lspMethodName,
-            languageServerName,
-            newParams,
-            cancellationToken).ConfigureAwait(false);
 
-        var result = csharpResults?.Response;
-        if (result is null)
+        foreach (var range in semanticTokensParams!.Ranges)
+        {
+            var newParams = new SemanticTokensRangeParams
+            {
+                TextDocument = semanticTokensParams.TextDocument,
+                PartialResultToken = semanticTokensParams.PartialResultToken,
+                Range = range,
+            };
+
+            var task = Task.Run(async () =>
+            {
+                using (var disposable = _telemetryReporter.TrackLspRequest(lspMethodName, languageServerName, semanticTokensParams.CorrelationId))
```

Comment:
>semanticTokensParams [](http://example.com/codeflow?start=110&length=20)

Seems like there could be some improvements here around what is captured by moving languageServerName/lspMethodName/newParams declarations inside Task.Run